readme.md
# 部署

#### 介绍

大商创 X 版本支持一键部署运行环境。[LNMP一键安装包](https://lnmp.org) 是一个用Linux Shell编写的可以为CentOS/RHEL/Fedora/Aliyun/Amazon、Debian/Ubuntu/Raspbian/Deepin/Mint Linux VPS或独立主机安装LNMP(Nginx/MySQL/PHP)、LNMPA(Nginx/MySQL/PHP/Apache)、LAMP(Apache/MySQL/PHP)生产环境的Shell程序。

#### 数据盘挂载（可选）

如有数据盘，请先挂载到 /ecmoban 目录，如数据盘已有数据，操作前请做好相应备份。参考步骤：https://help.aliyun.com/document_detail/25426.html

#### 环境准备

以 **CentOS** 为例，安装前确认已经安装基本工具

- 使用  命令安装 git、screen、wget 命令。 

```
yum install git screen wget -y
```

为防止掉线等情况，建议使用screen，执行：

```
screen -S lnmp
```

安装过程如断线可使用 `screen -r lnmp` 恢复。

#### 安装 LNMP 环境

1、执行LNMP安装命令

- 下载安装脚本

```
cd /usr/local/src && git clone https://gitee.com/dscmall/lnmp.git && cd lnmp && chmod -R +x ../lnmp/
```

- 执行安装脚本（注意 *PHPSelect* 参数的差异）

php 7.1

```
LNMP_Auto="y" DBSelect="4" DB_Root_Password="p12345678A" InstallInnodb="y" PHPSelect="7" SelectMalloc="1" ./install.sh lnmp
```

php 7.2

```
LNMP_Auto="y" DBSelect="4" DB_Root_Password="p12345678A" InstallInnodb="y" PHPSelect="8" SelectMalloc="1" ./install.sh lnmp
```

提示如下 说明安装完毕 

```
Install lnmp completed! enjoy it.
```

2、安装 swoole loader 扩展

执行安装脚本，确保当前位置是在 /usr/local/src/lnmp/ 下，然后执行如下命令 可以用 `pwd` 查看 当前所在目录

```
# php 7.1 扩展安装目录
cd ./loader/1.9 
# php 7.2 扩展安装目录
cd ./loader/2.2 
```

- 执行安装

```
./install.sh && /etc/init.d/nginx reload
```

会出现提示

```
Please select the php path to install loader extension:

1 : /usr/bin/php
2 : /usr/bin/php-fpm
```

选择 1 然后回车即可，安装完成之后，

- 检查安装状态

```
php --ri swoole_loader
```

安装成功后会出现如下提示

```
[root@centos dscmall]# php --ri swoole_loader

swoole_loader

swoole_loader support => enabled
swoole_loader version => 2.2.0
```

3、>>> **【重要安全提示】** <<< 请立即修改 mysql 的 root 用户密码

访问：[http://ip/pma](http://ip/pma)，使用 root 用户登录后，请立即修改初始密码（`p12345678A`）。

#### 其他

LNMP添加、删除虚拟主机及伪静态[使用教程](https://lnmp.org/faq/lnmp-vhost-add-howto.html)


ChangeLog
更新记录

LNMP 1.8

增加PHP 8.0支持；
增加Oracle Linux支持；
增加WSL优化支持；
增加CentOS6源自动调整；
增加BuyPass、ZeroSSL SSL证书；
增加php-memcache支持PHP 8.0；
增加imagick支持PHP 8.0；
增加apcu支持PHP 8.0；
优化时间同步；
优化Aliyunx Linux优化；
优化sudo下添加虚拟主机SSL；
优化CentOS8 Stream支持；
优化PHP下载；
更新诸多软件版本；
其他一些功能优化及调整......

LNMP 1.7

增加PHP 7.4支持；
增加MariaDB 10.4支持；
增加systemd service脚本；
增加phpmyadmin 5.0；
增加多PHP版本升级工具 ./upgrade.sh mphp，只支持7.2.x-7.2.x类似小版本升级，大版本直接新装即可；
移除MariaDB 10.0；
优化lnmp管理脚本；
优化数据库root密码重置工具；
优化系统版本获取脚本；
优化redis安装脚本；
优化fail2ban安装脚本；
优化nginx安装、升级脚本；
完善CentOS8、Ubuntu20、Deepin20、Fedora等新版本的支持；
完善MySQL升级时增加旧版本下载镜像；
调整php7 memcached使用稳定版；
修复lamp模式下ssl chain添加的bug；
更新诸多软件版本；
其他一些功能优化及调整......

LNMP 1.6

增加PHP 7.3支持；
增加MariaDB 10.3支持；
增加Nginx/Apache TLS 1.3支持；
增加使用DNS API方式只创建SSL证书不添加网站，命令lnmp onlyssl；
增加Fedora 30、Ubuntu 19.04、RHEL 8等新版本的支持；
增加nginx、apache等一些例子，并移至conf/example目录下；
增加在低于2048MB内存是增加2GB SWAP或低于1024MB时增加1GB SWAP的选项；
增加反向代理配置例子文件 nginx-reverse-proxy-example.conf；
增加只安装nginx模式增加拷贝lnmp管理工具；
增加 RHELRepo=local 参数，RHEL安装时使用RHEL本地源，不设置源为163 centos源；
增加 CheckMirror=n 参考，安装时不检查下载镜像，方便无网络安装；
优化不安装数据库时的处理流程；
优化acme.sh规则可能不生效的情况；
优化nginx系统满足条件是启用reuseport；
优化程序代码目录清理；
优化重写nginx管理脚本；
优化安装数据库时的内存判断，放宽内存限制；
修复新版深度Deepin 15.8桌面版问题；
修复新版Fedora 29下问题；
调整mysql使用带boost源码，同时支持不带boost安装方式；
调整卸载时不删除安装时添加的iptables规则；
调整fileinfo安装选项，内存满足时自动安装(可关闭)；
升级Nginx至1.16.0；
升级各程序版本；
更新诸多软件版本；
其他一些功能优化及调整......

LNMP 1.5

增加PHP 7.2支持；
增加MariaDB 10.2支持；
增加MySQL 8.0支持；
增加PHP模块fileinfo选项，安装或升级时可以选择是否安装开启，选项在安装包目录下 lnmp.conf 中 Enable_PHP_Fileinfo 默认为 n 不开启，如有安装设置为 y ；
增加Nginx lua模块选项，安装或升级时可以选择是否为nginx安装开启lua，选项在安装包目录下 lnmp.conf 中 Enable_Nginx_Lua 默认为 n 不开启，如有安装设置为 y ；
增加thinkphp、yii2、codeigniter、laravel伪静态；
增加使用acme.sh生成SSL证书，移除官方的certbot；
增加Nginx编译参数，默认开启stream；
增加Apache 2.4支持http2；
增加旧版本Ubuntu中国地区源；
增加虚拟主机添加时pathinfo可选项；
增加certbot升级到acme.sh的一键升级功能命令为 ./upgrade1.x-1.5.sh ssl 按提示输入和原来SSL虚拟主机一样配置的信息就可以；
增加多核编译加快安装时间；
增加kali Linux发行版支持；
增加对Amazone Linux 2支持；
增加修复MySQL gcc7 bug；
增加magento2实例配置文件 magento2-example.conf；
增加无人值守安装，官网上有具体教程；
增加LetseEcrypt 通配符/泛域名 SSL证书生成，官网上有具体教程；
优化MySQL/MariaDB root密码重装工具；
修复深度Deepin桌面版问题；
修复MySQL bug；
修复MariaDB bug；
移除PHP禁用函数scandir；
调整默认PHP版本为5.6；
调整默认Apache版本为2.4；
调整MySQL/MariaDB安装时未输入密码由原来密码设为 root 调整为 lnmp.org#随机数字 ；
调整MySQL国内下载源为科大的；
调整LAMP模式下SSL添加过程增加中间证书可选项；
调整默认虚拟主机页面字体显示等细节；
调整nginx伪静态至/usr/local/nginx/conf/rewrite目录；
调整优化lnmp下载源选择；
完善卸载脚本；
完善优化lnmp升级脚本 upgrade1.x-1.5.sh；
调整数据库内存优化加载方法；
其他一些功能优化及调整......

LNMP 1.4

增加LNMP模式下多PHP版本安装，安装包目录下运行：./install.sh mphp 进行安装；
增加单独nginx安装，安装包目录下运行：./install.sh nginx 进行安装；
增加单独数据库安装，安装包目录下运行：./install.sh db 进行安装；
增加Let's Encrypt自动生成证书和自备SSL证书支持；
增加不安装数据库选项；
增加多PHP版本下扩展安装支持；
增加ioncube php7的支持；
增加apcu php扩展；
增加PHP7.1支持；
增加lnmp管理脚本升级脚本，安装包目录下运行：./upgrade1.x-1.4.sh；
增加对中国用户的识别，使用国内源；
增加Amazon Linux、Linux Mint支持；
增加计算显示LNMP安装使用时间；
增加是否已安装LNMP判断，防止重复安装覆盖数据；
增加在已存在目录下安装数据库时先进行备份；
增加防跨目录配置文件移除脚本，安装包目录里tools目录下运行 ./remove_open_basedir_restriction.sh ；
lnmp管理脚本增加SSL选项，可以选择Let's Encrypt或自定义证书和key文件，Letsencrypt每7天自动尝试续期；
lnmp管理脚本增加ftp账号信息显示；
lnmp管理脚本增加LNMP模式下添加虚拟主机时多PHP版本选择选项；
lnmp管理脚本去除add more domain name提示，添加直接输入域名，不添加直接回车；
lnmp管理脚本添加虚拟主机处理流程上的优化；
lnmp管理脚本增加多PHP状态管理支持；
lnmp管理脚本增加 lnmp ftp show 显示ftp账号信息；
lnmp管理脚本已存在虚拟主机SSL时可以直接lnmp ssl add 进行添加；
nginx默认使用openssl 1.0.2进行安装，以便更好的支持http2；
修复完善国内composer安装问题，国内能正常流畅的安装；
修复lnmpa/lamp模式下https问题；
修复arm下centos不存在libcurl依赖包的问题；
调整部分软件源码包格式为tar.bz2；
调整apache日志LogLevel为crit；
调整php模块的引入方式,使用with-config-file-scan-dir引入；
调整默认PHP版本调整为PHP 5.5；
调整国内外节点顺序，1,3,5 为国内节点(soft1.vpser.net 可调整数字)，2,4,6,8 为国外节点(soft2.vpser.net 可调整数字)。
调整默认虚拟主机不在使用www.lnmp.org域名，使用_代替；
增加lnmp.conf配置文件中nginx是否使用新版openssl编译安装选项；
增加LNMP模式下上传目录或不需php解析目录php禁止访问的实例代码，去掉注释修改其中目录重启nginx即可使用；
增加LNMP模式下的https实例配置文件：/usr/local/nginx/conf/enable-ssl-example.conf；
增加LNMPA模式下的https实例配置文件：/usr/local/nginx/conf/enable-lnmpa-ssl-vhost-example.conf；
增加LAMP模式下的https实例配置文件：/usr/local/apache/conf/enable-apache-ssl-vhost-example.conf；
增加PHP模块zts和non-zts自动识别；
添加Ubuntu 17.04支持；
php探针在php7下优化；
fastcgi.conf增加防跨目录的二次设置，移除参考：https://lnmp.org/faq/lnmp-vhost-add-howto.html#user.ini ；
imagemagick增加webp格式支持；
去掉pureftp配置文件匿名模式下匿名用户上传权限；
防止误卸载，卸载脚本增加备份数据库目录；
优化pureftpd安装脚本，可单独在服务器进行安装；
优化nginx虚拟主机配置文件，增加对let'sencrypt目录的允许规则；
优化卸载脚本；
优化组件模块安装脚本；
优化部分启动脚本；
调整升级MySQL/MariaDB时my.cnf和启动脚本的位置到MySQL/MariaDB备份目录下；
调整优化fail2ban、denyhosts安装脚本；
诸多小的优化及调整......

LNMP 1.3

优化安装脚本、升级脚本及组件安装脚本；
增加lnmp.conf配置文件，可以自定义下载服务器地址、数据库目录及nginx和php编译参数；不论安装升级都会调用该文件里的设置；
增加安装和升级时会根据配置自动优化相关参数；
增加php7支持；
增加mysql 5.7支持；
增加phpmyadmin升级功能；
php增加Composer工具；
php 5.4以上版本增加intl、xsl扩展；
增加nginx升级到1.9.5以上版本支持http2；
增加php-redis、php-memcached、php-memcache、imagick PHP7下模块的支持；
增加升级PHP/MySQL/MariaDB时对源目录进行完整备份，防范升级失败风险；
增加卸载LNMP/LNMPA/LAMP时对数据库目录进行备份功能；
增加lnmp脚本创建虚拟主机时可选创建MySQL数据库和FTP账号选项(需先安装好pureftpd)；
增加安装或升级完清理目录节约硬盘空间；
redis仅保留最新版，删除旧版支持；
修正在部分系统下MySQL无法正确设置密码的问题；
修正完善防火墙规则；
pureftpd去掉MySQL的依赖和web面板，只使用lnmp ftp {add|list|edit|del} 进行ftp账号管理；
调整pureftpd配置文件目录至/usr/local/pureftpd/etc/下；
nginx.conf php启用配置简化为include enable-php.conf;
诸多小的更新及调整......


2015年6月1日发布LNMP 1.2

完全重写优化安装脚本、升级脚本及组件安装等大部分脚本；
状态管理、虚拟主机管理、FTP管理及数据库管理全部采用lnmp命令替换，去掉/root/vhost.sh和/root/lnmp脚本；
Nginx、MySQL/MariaDB、PHP的升级脚本整合为upgrade.sh;
缓存、加速、imageMagick、ionCube等的安装卸载采用addons.sh进行；
增加SSH防护工具DenyHosts、Failban安装脚本及DenyHosts被封IP删除脚本；
增加新的备份脚本backup.sh；
支持的Linux发行版：CentOS 5-7、RHEL 6-7、Fedora 20-21、Aliyun、Debian 6-8、Raspbian、Ubuntu 10.10-15.04；
支持ARM平台，Raspberry Pi已测试；
MySQL安装时可选5.1.73、5.5.42、5.6.23，MariaDB可选5.5.42、10.0.17；
PHP安装时可选5.2.17、5.3.29、5.4.41、5.5.25、5.6.9；
安装时增加内存优化jemalloc或gperftools；
Apache安装时可选2.2.29或2.4.12；
Apache日志统一到/ecmoban/wwwlogs/下；
更新Nginx至1.8.0；
更新Zend Guard Loader至支持PHP 5.6；
Nginx配置文件进行了调整，启用php和pathinfo采用include方式，增加ssl范例配置；



2014年6月1日发布LNMP 1.1

调整优化脚本安装脚本；
安装时增加MariaDB选项；
增加MySQL升级脚本 upgrade_mysql.sh；
增加MySQL升级到MariaDB脚本 upgrade_mysql2mariadb.sh；
增加MariaDB升级脚本 upgrade_mariadb.sh；
增加Redis安装脚本 redis.sh；
增加OPcache安装脚本 opcache.sh；
增加MySQL/MariaDB root密码重置脚本；
添加pathinfo支持，去掉include pathinfo的#，在try_files前加#重启nginx即可使用；
重写nginx init.d脚本；
优化PHP升级脚本；

upgrade1.x-1.7.sh
#!/usr/bin/env bash
export PATH=$PATH:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:~/bin

# Check if user is root
if [ $(id -u) != "0" ]; then
    echo "Error: You must be root to run this script"
    exit 1
fi

cur_dir=$(pwd)
isSSL=$1

. lnmp.conf
. include/main.sh

Get_Dist_Name
Check_Stack
Check_DB

Upgrade_Dependent()
{
    if [ "$PM" = "yum" ]; then
        Echo_Blue "[+] Yum installing dependent packages..."
        Get_Dist_Version
        for packages in patch wget crontabs unzip tar ca-certificates net-tools libc-client-devel psmisc libXpm-devel git-core c-ares-devel libicu-devel libxslt libxslt-devel xz expat-devel bzip2 bzip2-devel libaio-devel rpcgen libtirpc-devel perl python-devel cyrus-sasl-devel sqlite-devel oniguruma-devel re2c;
        do yum -y install $packages; done
        yum -y update nss

        if [ "${DISTRO}" = "CentOS" ] && echo "${CentOS_Version}" | grep -Eqi "^8"; then
            if ! yum repolist all|grep PowerTools; then
                echo "PowerTools repository not found, add PowerTools repository ..."
                cat >/etc/yum.repos.d/CentOS-PowerTools.repo<<EOF
[PowerTools]
name=CentOS-\$releasever - PowerTools
mirrorlist=http://mirrorlist.centos.org/?release=\$releasever&arch=\$basearch&repo=PowerTools&infra=\$infra
#baseurl=http://mirror.centos.org/\$contentdir/\$releasever/PowerTools/\$basearch/os/
gpgcheck=1
enabled=0
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-centosofficial
EOF
            fi
            dnf --enablerepo=PowerTools install rpcgen re2c -y
            dnf --enablerepo=PowerTools install oniguruma-devel -y
        fi

        if echo "${CentOS_Version}" | grep -Eqi "^7" || echo "${RHEL_Version}" | grep -Eqi "^7"; then
            yum -y install epel-release
            if [ "${country}" = "CN" ]; then
                sed -i "s@^#baseurl=http://download.fedoraproject.org/pub@baseurl=http://mirrors.aliyun.com@g" /etc/yum.repos.d/epel*.repo
                sed -i "s@^metalink@#metalink@g" /etc/yum.repos.d/epel*.repo
            fi
            yum -y install oniguruma oniguruma-devel
            if [ "${CheckMirror}" = "n" ]; then
                cd ${cur_dir}/src/
                yum -y install ./oniguruma-6.8.2-1.el7.x86_64.rpm
                yum -y install ./oniguruma-devel-6.8.2-1.el7.x86_64.rpm
            fi
        fi
    elif [ "$PM" = "apt" ]; then
        Echo_Blue "[+] apt-get installing dependent packages..."
        apt-get update -y
        for packages in debian-keyring debian-archive-keyring build-essential bison libkrb5-dev libcurl3-gnutls libcurl4-gnutls-dev libcurl4-openssl-dev libcap-dev ca-certificates libc-client2007e-dev psmisc patch git libc-ares-dev libicu-dev e2fsprogs libxslt1.1 libxslt1-dev libc-client-dev xz-utils libexpat1-dev bzip2 libbz2-dev libaio-dev libtirpc-dev python-dev libsqlite3-dev libonig-dev;
        do apt-get --no-install-recommends install -y $packages; done
    fi
}

if [ "${isSSL}" == "ssl" ]; then
    echo "+--------------------------------------------------+"
    echo "|  A tool to upgrade lnmp 1.4 certbot to acme.sh   |"
    echo "+--------------------------------------------------+"
    echo "|For more information please visit https://lnmp.org|"
    echo "+--------------------------------------------------+"
    if [[ "${Get_Stack}" =~ "lnmp" ]]; then
        domain=""
        while :;do
            Echo_Yellow "Please enter domain(example: www.lnmp.org): "
            read domain
            if [ "${domain}" != "" ]; then
                if [ ! -f "/usr/local/nginx/conf/vhost/${domain}.conf" ]; then
                    Echo_Red "${domain} is not exist,please check!"
                    exit 1
                else
                    echo " Your domain: ${domain}"
                    if ! grep -q "/etc/letsencrypt/live/${domain}/fullchain.pem" "/usr/local/nginx/conf/vhost/${domain}.conf"; then
                        Echo_Red "SSL configuration NOT found in the ${domain} config file!"
                        exit 1
                    fi
                    break
                fi
            else
                Echo_Red "Domain name can't be empty!"
            fi
        done

        Echo_Yellow "Enter more domain name(example: lnmp.org *.lnmp.org): "
        read moredomain
        if [ "${moredomain}" != "" ]; then
            echo " domain list: ${moredomain}"
        fi

        vhostdir="/ecmoban/wwwroot/${domain}"
        echo "Please enter the directory for the domain: $domain"
        Echo_Yellow "Default directory: /ecmoban/wwwroot/${domain}: "
        read vhostdir
        if [ "${vhostdir}" == "" ]; then
            vhostdir="/ecmoban/wwwroot/${domain}"
        fi
        echo "Virtual Host Directory: ${vhostdir}"

        if [ ! -d "${vhostdir}" ]; then
            Echo_Red "${vhostdir} does not exist or is not a directory!"
            exit 1
        fi

        letsdomain=""
        if [ "${moredomain}" != "" ]; then
            letsdomain="-d ${domain}"
            for i in ${moredomain};do
                letsdomain=${letsdomain}" -d ${i}"
            done
        else
            letsdomain="-d ${domain}"
        fi

        if [ -s /usr/local/acme.sh/acme.sh ]; then
            echo "/usr/local/acme.sh/acme.sh [found]"
        else
            cd /tmp
            [[ -f latest.tar.gz ]] && rm -f latest.tar.gz
            wget https://soft.vpser.net/lib/acme.sh/latest.tar.gz --prefer-family=IPv4 --no-check-certificate
            tar zxf latest.tar.gz
            cd acme.sh-*
            ./acme.sh --install --log --home /usr/local/acme.sh --certhome /usr/local/nginx/conf/ssl
            cd ..
            rm -f latest.tar.gz
            rm -rf acme.sh-*
            sed -i 's/cat "\$CERT_PATH"$/#cat "\$CERT_PATH"/g' /usr/local/acme.sh/acme.sh
            if command -v yum >/dev/null 2>&1; then
                yum -y update nss
                service crond restart
                chkconfig crond on
            elif command -v apt-get >/dev/null 2>&1; then
                /etc/init.d/cron restart
                update-rc.d cron defaults
            fi
        fi

        . "/usr/local/acme.sh/acme.sh.env"

        if [ -s /usr/local/nginx/conf/ssl/${domain}/fullchain.cer ]; then
            echo "Removing exist domain certificate..."
            rm -rf /usr/local/nginx/conf/ssl/${domain}
        fi

        echo "Starting create SSL Certificate use Let's Encrypt..."
        /usr/local/acme.sh/acme.sh --issue ${letsdomain} -w ${vhostdir} --reloadcmd "/etc/init.d/nginx reload"
        lets_status=$?
        if [ "${lets_status}" = 0 ]; then
            Echo_Green "Let's Encrypt SSL Certificate create successfully."
            echo "Modify ${domain} configure..."
            sed -i "s@/etc/letsencrypt/live/${domain}/fullchain.pem@/usr/local/nginx/conf/ssl/${domain}/fullchain.cer@g" "/usr/local/nginx/conf/vhost/${domain}.conf"
            sed -i "s@/etc/letsencrypt/live/${domain}/privkey.pem@/usr/local/nginx/conf/ssl/${domain}/${domain}.key@g" "/usr/local/nginx/conf/vhost/${domain}.conf"
            echo "done."

            if crontab -l|grep -q "/bin/certbot renew"; then
                (crontab -l | grep -v "/bin/certbot renew") | crontab -
            fi

            /etc/init.d/nginx reload
            sleep 1
            Echo_Green "upgrade ${domain} successfully."
        else
            Echo_Red "Let's Encrypt SSL Certificate create failed!"
            Echo_Red "upgrade ${domain} fialed."
        fi
    elif [ "${Get_Stack}" == "lamp" ]; then
        domain=""
        while :;do
            Echo_Yellow "Please enter domain(example: www.lnmp.org): "
            read domain
            if [ "${domain}" != "" ]; then
                if [ ! -f "/usr/local/apache/conf/vhost/${domain}.conf" ]; then
                    Echo_Red "${domain} is not exist,please check!"
                    exit 1
                else
                    echo " Your domain: ${domain}"
                    if ! grep -q "/etc/letsencrypt/live/${domain}/privkey.pem" "/usr/local/apache/conf/vhost/${domain}.conf"; then
                        Echo_Red "SSL configuration NOT found in the ${domain} config file!"
                        exit 1
                    fi
                    break
                fi
            else
                Echo_Red "Domain name can't be empty!"
            fi
        done

        Echo_Yellow "Enter more domain name(example: lnmp.org *.lnmp.org): "
        read moredomain
        if [ "${moredomain}" != "" ]; then
            echo " domain list: ${moredomain}"
        fi

        vhostdir="/ecmoban/wwwroot/${domain}"
        echo "Please enter the directory for the domain: $domain"
        Echo_Yellow "Default directory: /ecmoban/wwwroot/${domain}: "
        read vhostdir
        if [ "${vhostdir}" == "" ]; then
            vhostdir="/ecmoban/wwwroot/${domain}"
        fi
        echo "Virtual Host Directory: ${vhostdir}"

        if [ ! -d "${vhostdir}" ]; then
            Echo_Red "${vhostdir} does not exist or is not a directory!"
            exit 1
        fi

        letsdomain=""
        if [ "${moredomain}" != "" ]; then
            letsdomain="-d ${domain}"
            for i in ${moredomain};do
                letsdomain=${letsdomain}" -d ${i}"
            done
        else
            letsdomain="-d ${domain}"
        fi

        if [ -s /usr/local/acme.sh/acme.sh ]; then
            echo "/usr/local/acme.sh/acme.sh [found]"
        else
            cd /tmp
            [[ -s latest.tar.gz ]] && rm -f latest.tar.gz
            wget https://soft.vpser.net/lib/acme.sh/latest.tar.gz --prefer-family=IPv4 --no-check-certificate
            tar zxf latest.tar.gz
            cd acme.sh-*
            ./acme.sh --install --log --home /usr/local/acme.sh --certhome /usr/local/apache/conf/ssl
            cd ..
            rm -f latest.tar.gz
            rm -rf acme.sh-*
            sed -i 's/cat "\$CERT_PATH"$/#cat "\$CERT_PATH"/g' /usr/local/acme.sh/acme.sh
            if command -v yum >/dev/null 2>&1; then
                yum -y update nss
                yum -y install ca-certificates
                service crond restart
                chkconfig crond on
            elif command -v apt-get >/dev/null 2>&1; then
                /etc/init.d/cron restart
                update-rc.d cron defaults
            fi
        fi

        . "/usr/local/acme.sh/acme.sh.env"

        if [ -s /usr/local/apache/conf/ssl/${domain}/fullchain.cer ]; then
            echo "Removing exist domain certificate..."
            rm -rf /usr/local/apache/conf/ssl/${domain}
        fi

        echo "Starting create SSL Certificate use Let's Encrypt..."
        /usr/local/acme.sh/acme.sh --issue ${letsdomain} -w ${vhostdir} --reloadcmd "/etc/init.d/httpd graceful"
        lets_status=$?
        if [ "${lets_status}" = 0 ]; then
            Echo_Green "Let's Encrypt SSL Certificate create successfully."
            echo "Modify ${domain} configure..."
            sed -i "s@/etc/letsencrypt/live/${domain}/fullchain.pem@/usr/local/apache/conf/ssl/${domain}/${domain}.cer@g" "/usr/local/apache/conf/vhost/${domain}.conf"
            sed -i "s@/etc/letsencrypt/live/${domain}/privkey.pem@/usr/local/apache/conf/ssl/${domain}/${domain}.key@g" "/usr/local/apache/conf/vhost/${domain}.conf"
            sed -i "/\/usr\/local\/apache\/conf\/ssl\/${domain}\/${domain}.key/a\SSLCertificateChainFile \/usr\/local\/apache\/conf\/ssl\/${domain}\/ca.cer" "/usr/local/apache/conf/vhost/${domain}.conf"
            echo "done."

            if crontab -l|grep -q "/bin/certbot renew"; then
                (crontab -l | grep -v "/bin/certbot renew") | crontab -
            fi

            /etc/init.d/httpd graceful
            sleep 1
            Echo_Green "upgrade ${domain} successfully."
        else
            Echo_Red "Let's Encrypt SSL Certificate create failed!"
            Echo_Red "upgrade ${domain} fialed."
        fi

    else
        Echo_Red "Can't get stack info and will not be able to upgrade."
    fi
else
    echo "+--------------------------------------------------+"
    echo "|  A tool to upgrade lnmp manager from 1.x to 1.7  |"
    echo "+--------------------------------------------------+"
    echo "|For more information please visit https://lnmp.org|"
    echo "+--------------------------------------------------+"
    Press_Start
    if [ "${Get_Stack}" == "unknow" ]; then
        Echo_Red "Can't get stack info."
        exit
    elif [ "${Get_Stack}" == "lnmp" ]; then
        Upgrade_Dependent
        echo "Copy lnmp manager..."
        sleep 1
        \cp ${cur_dir}/conf/lnmp /bin/lnmp
        chmod +x /bin/lnmp
        echo "Copy configure files..."
        sleep 1
        if [ ! -s /usr/local/nginx/conf/enable-php.conf ]; then
            \cp ${cur_dir}/conf/enable-php.conf /usr/local/nginx/conf/enable-php.conf
        fi
        if [ ! -s /usr/local/nginx/conf/pathinfo.conf ]; then
            \cp ${cur_dir}/conf/pathinfo.conf /usr/local/nginx/conf/pathinfo.conf
        fi
        if [ ! -s /usr/local/nginx/conf/enable-php-pathinfo.conf ]; then
            \cp ${cur_dir}/conf/enable-php-pathinfo.conf /usr/local/nginx/conf/enable-php-pathinfo.conf
        fi
        if [ ! -d /usr/local/nginx/conf/rewrite ]; then
            \cp -ra ${cur_dir}/conf/rewrite /usr/local/nginx/conf/
        fi
        if [ ! -d /usr/local/nginx/conf/vhost ]; then
            mkdir /usr/local/nginx/conf/vhost
        fi
    elif [ "${Get_Stack}" == "lnmpa" ]; then
        Upgrade_Dependent
        echo "Copy lnmp manager..."
        sleep 1
        \cp ${cur_dir}/conf/lnmpa /bin/lnmp
        chmod +x /bin/lnmp
        echo "Copy configure files..."
        sleep 1
        \cp ${cur_dir}/conf/proxy.conf /usr/local/nginx/conf/proxy.conf
        if [ ! -s /usr/local/nginx/conf/proxy-pass-php.conf ]; then
            \cp ${cur_dir}/conf/proxy-pass-php.conf /usr/local/nginx/conf/proxy-pass-php.conf
        fi
        if ! grep -q "SetEnvIf X-Forwarded-Proto https HTTPS=on" /usr/local/apache/conf/httpd.conf; then
            if /usr/local/apache/bin/httpd -v|grep -Eqi "Apache/2.2."; then
                sed -i "/Include conf\/vhost\/\*.conf/i\SetEnvIf X-Forwarded-Proto https HTTPS=on\n" /usr/local/apache/conf/httpd.conf
            elif /usr/local/apache/bin/httpd -v|grep -Eqi "Apache/2.4."; then
                sed -i "/IncludeOptional conf\/vhost\/\*.conf/i\SetEnvIf X-Forwarded-Proto https HTTPS=on\n" /usr/local/apache/conf/httpd.conf
            fi
        fi
        if [ ! -d /usr/local/nginx/conf/vhost ]; then
            mkdir /usr/local/nginx/conf/vhost
        fi
    elif [ "${Get_Stack}" == "lamp" ]; then
        Upgrade_Dependent
        echo "Copy configure files..."
        sleep 1
        \cp ${cur_dir}/conf/lamp /bin/lnmp
        chmod +x /bin/lnmp
        echo "Copy configure files..."
        sleep 1
        if /usr/local/apache/bin/httpd -v|grep -Eqi "Apache/2.2."; then
            \cp ${cur_dir}/conf/httpd22-ssl.conf  /usr/local/apache/conf/extra/httpd-ssl.conf
        elif /usr/local/apache/bin/httpd -v|grep -Eqi "Apache/2.4."; then
            \cp ${cur_dir}/conf/httpd24-ssl.conf  /usr/local/apache/conf/extra/httpd-ssl.conf
            sed -i 's/^#LoadModule socache_shmcb_module/LoadModule socache_shmcb_module/g' /usr/local/apache/conf/httpd.conf
            sed -i 's/^LoadModule lbmethod_heartbeat_module/#LoadModule lbmethod_heartbeat_module/g' /usr/local/apache/conf/httpd.conf
        fi
        if [ ! -d /usr/local/apache/conf/vhost ]; then
            mkdir /usr/local/apache/conf/vhost
        fi
    fi

    if [ "${DB_Name}" = "mariadb" ]; then
        sed -i 's#/etc/init.d/mysql#/etc/init.d/mariadb#' /bin/lnmp
    elif [ "${DB_Name}" = "None" ]; then
        sed -i 's#/etc/init.d/mysql.*##' /bin/lnmp
    fi

    if [ -s /usr/local/acme.sh/acme.sh ]; then
        /usr/local/acme.sh/acme.sh --upgrade
        sed -i 's/cat "\$CERT_PATH"$/#cat "\$CERT_PATH"/g' /usr/local/acme.sh/acme.sh
    fi

    Echo_Green "upgrade lnmp manager complete."
fi